metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -0,0 +1,235 @@
+import scala.meta.internal.metals.DebugUnresolvedMainClassParams
+import scala.meta.internal.metals.DebugUnresolvedTestClassParams
+import scala.meta.internal.metals.BuildTargetClassesFinder
+import scala.meta.internal.metals.Compilations
+import scala.util.Try
+import scala.util.Failure
+import org.eclipse.lsp4j.MessageParams
+import org.eclipse.lsp4j.MessageType
+import scala.meta.internal.metals.JsonParser
+import scala.meta.internal.metals.JsonParser._
+import scala.meta.internal.metals.ClassNotFoundInBuildTargetException
+import scala.meta.internal.metals.Messages.UnresolvedDebugSessionParams
+
+class DebugProvider(
+    definitionProvider: DefinitionProvider,
+    buildServer: => Option[BuildServerConnection],
+    buildTargets: BuildTargets,
+    buildTargetClasses: BuildTargetClasses,
+    compilations: Compilations,
+    languageClient: MetalsLanguageClient
+) {
+
+  lazy val buildTargetClassesFinder = new BuildTargetClassesFinder(
+    buildTargets,
+    buildTargetClasses
+  )
+
+  def start(
+      parameters: b.DebugSessionParams
+  )(implicit ec: ExecutionContext): Future[DebugServer] = {
+    Future.fromTry(parseSessionName(parameters)).flatMap { sessionName =>
+      val proxyServer = new ServerSocket(0)
+      val host = InetAddresses.toUriString(proxyServer.getInetAddress)
+      val port = proxyServer.getLocalPort
+      val uri = URI.create(s"tcp://$host:$port")
+      val connectedToServer = Promise[Unit]()
+
+      val awaitClient =
+        () => Future(proxyServer.accept()).withTimeout(10, TimeUnit.SECONDS)
+
+      // long timeout, since server might take a while to compile the project
+      val connectToServer = () => {
+        buildServer
+          .map(_.startDebugSession(parameters))
+          .getOrElse(BuildServerUnavailableError)
+          .withTimeout(60, TimeUnit.SECONDS)
+          .map { uri =>
+            val socket = connect(uri)
+            connectedToServer.trySuccess(())
+            socket
+          }
+      }
+
+      val proxyFactory = { () =>
+        val targets = parameters.getTargets.asScala
+          .map(_.getUri)
+          .map(new BuildTargetIdentifier(_))
+        val sourcePathProvider = new SourcePathProvider(
+          definitionProvider,
+          buildTargets,
+          targets.toList
+        )
+        DebugProxy
+          .open(sessionName, sourcePathProvider, awaitClient, connectToServer)
+      }
+      val server = new DebugServer(sessionName, uri, proxyFactory)
+
+      server.listen.andThen { case _ => proxyServer.close() }
+
+      connectedToServer.future.map(_ => server)
+    }
+  }
+
+  def resolveMainClassParams(
+      params: DebugUnresolvedMainClassParams
+  )(implicit ec: ExecutionContext): Future[b.DebugSessionParams] = {
+    withRebuildRetry(() =>
+      buildTargetClassesFinder
+        .findMainClassAndItsBuildTarget(
+          params.mainClass,
+          Option(params.buildTarget)
+        )
+    ).map {
+      case (clazz, target) :: others =>
```

Comment:
This will also fail on Nil, could you handle that case?

Reply:
as I commented on `BuildTargetClassFinder#findMainClassAndItsBuildTarget` (above method def), it either returns `Success` with non-empty list, or `Failure`. So it should not be the case. I wish we had a type-level solution for that.

Reply:
Solution for NonEmptyList? You have `scalaz.NonEmptyList`, `cats.data.NonEmptyList`, everything is already included in dependencies

Reply:
hmm, last time I checked (several months ago), `cats` just got eliminated from Metals, and it was said it's better to have less dependencies. Okay, thanks, @kpbochenek !

Reply:
okay, seems like there's no cats used in the main code, and @tgodzik recommended not to start.
^ fixed with `throw NoSuchElement(Exception)` for the `Nil` case